<a href="https://colab.research.google.com/github/0zdata04/Myacount-/blob/main/Business_Intelligence_SuperTienda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [45]:
%load_ext sql
%config SqlMagic.style = '_DEPRECATED_DEFAULT'
%sql sqlite:////34_SuperTienda_Espanol.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


#1.Objetivo del Datase



El Dataset Maestro se construye con el propósito de centralizar la información comercial más relevante de SuperTienda en una única estructura de datos. Esta integración permitirá responder las principales preguntas de negocio relacionadas con el comportamiento de las ventas, los clientes, los productos y la distribución geográfica.

Al consolidar esta información en un solo dataset, se facilitará el análisis posterior y se reducirá la necesidad de realizar múltiples consultas sobre la base de datos original.

##2.**Tabla de columnas**

| Columna         | Tabla Origen   | Justificación de negocio                                                                                    |
| --------------- | -------------- | ----------------------------------------------------------------------------------------------------------- |
| Fecha_Pedido    | Pedido         | Permite analizar la evolución temporal de las ventas e identificar tendencias comerciales.                  |
| ID_Pedido       | Pedido         | Identifica cada transacción comercial y permite medir el volumen de pedidos realizados.                     |
| ID_Cliente      | Cliente        | Permite identificar a cada cliente para analizar su comportamiento de compra.                               |
| Nombre_Cliente  | Cliente        | Facilita la interpretación de los resultados por parte del negocio.                                         |
| Segmento        | Cliente        | Permite comparar el comportamiento comercial entre diferentes tipos de clientes.                            |
| Ciudad          | Geografía      | Permite identificar las ciudades con mayor y menor desempeño comercial.                                     |
| Región          | Geografía      | Facilita el análisis geográfico por regiones y la comparación territorial.                                  |
| ID_Producto     | Producto       | Identifica de forma única cada producto vendido.                                                            |
| Nombre_Producto | Producto       | Permite reconocer fácilmente los productos dentro de los análisis.                                          |
| Categoría       | Producto       | Facilita el análisis de rentabilidad por líneas de productos.                                               |
| Subcategoría    | Producto       | Permite realizar análisis más específicos dentro de cada categoría.                                         |
| Cantidad        | Detalle_Pedido | Permite medir el volumen de productos comercializados.                                                      |
| Ventas          | Detalle_Pedido | Representa el ingreso generado por cada operación comercial y constituye la principal métrica del proyecto. |
| Ganancia        | Detalle_Pedido | Permite evaluar la rentabilidad del negocio y de los productos comercializados.                             |
| Descuento       | Detalle_Pedido | Permite analizar el impacto de las políticas de descuento sobre las ventas y la rentabilidad.               |


##3.Preguntas de negocio

###Nuestro Dataset deberá responder:

1. ¿Cómo evolucionaron las ventas por mes?

2. ¿Cuáles son las ciudades con mayores ventas?

3. ¿Qué productos generan más ganancias?

4. ¿Qué clientes generan mayores ingresos?

5. ¿Cuáles son las categorías más rentables?

##4.**Selección de tablas**

###Tabla: **Detalle_Pedido**

Razón de negocio

Fue seleccionada como tabla principal porque contiene las métricas fundamentales del negocio: ventas, cantidad, ganancias y descuentos. Estas variables permiten responder la mayoría de las preguntas planteadas en el proyecto.


---


###Tabla: **Pedido**

Razón de negocio

Se incorpora para agregar el contexto temporal de cada venta mediante la fecha del pedido y para relacionar las ventas con los clientes.


---


###Tabla: **Cliente**

Razón de negocio

Permite identificar quién realizó la compra y analizar el comportamiento comercial de los distintos segmentos de clientes.


---


###Tabla: **Cliente_Ubicación**

Razón de negocio

Actúa como tabla de relación entre los clientes y la información geográfica, permitiendo asociar correctamente cada cliente con su ubicación.


---


###Tabla: **Geografía**

Razón de negocio

Permite analizar el desempeño comercial por ciudad y región, identificando diferencias geográficas en las ventas.


---


###Tabla: **Producto**

Razón de negocio

Proporciona la información necesaria para evaluar el rendimiento comercial y la rentabilidad de cada producto, categoría y subcategoría.



##5.**Construcion del Dataset Maestro**

In [48]:
%%sql
SELECT
-- Información temporal
p.Fecha_Pedido,
p.ID_Pedido,

-- Información cliente
c.ID_Cliente,
c.Nombre_Cliente,
c.Segmento,

-- Información geográfica
g.ciudad,
g.region,

-- Información producto
pr.ID_producto,
pr.Categoria,
pr.Subcategoria,
pr.Nombre_Producto,

-- Métricas comerciales
dp.Ventas,
dp.Ganancia,
dp.Cantidad,
dp.Descuento

FROM "Detalle_Pedido" AS dp

-- INNER JOIN 1
INNER JOIN Pedidos as p
  ON dp.ID_Pedido = p.ID_Pedido

-- INNER JOIN 2
INNER JOIN Clientes as c
  ON p.ID_Cliente = c.ID_Cliente

--INNER JOIN 3
INNER JOIN Cliente_Ubicacion as  cu
  ON c.ID_Cliente = cu.ID_Cliente

-- INNER JOIN 4
INNER JOIN Geografia as g
  ON cu.ID_Ubicacion = g.ID_Ubicacion

-- INNER JOIN 5
INNER JOIN Productos as  pr
  ON  dp.ID_Producto = pr.ID_Producto

 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


Fecha_Pedido,ID_Pedido,ID_Cliente,Nombre_Cliente,Segmento,Ciudad,Region,ID_Producto,Categoria,Subcategoria,Nombre_Producto,Ventas,Ganancia,Cantidad,Descuento
2025-03-31,ORD-2023-00001,C0174,Nicolás Martínez,Corporativo,Puerto Montt,Sur,P0082,Oficina,Papelería,Papelería Eco 2,590.86,322.42,1,0.0
2025-03-31,ORD-2023-00001,C0174,Nicolás Martínez,Corporativo,Puerto Montt,Sur,P0119,Oficina,Sobres,Sobres Plus 9,4341.68,1907.85,7,0.0
2025-03-31,ORD-2023-00001,C0174,Nicolás Martínez,Corporativo,Puerto Montt,Sur,P0048,Muebles,Sillas,Sillas Plus 8,1250.86,281.32,2,0.0
2025-03-31,ORD-2023-00001,C0174,Nicolás Martínez,Corporativo,Puerto Montt,Sur,P0079,Muebles,Almacenamiento,Almacenamiento Nova 9,244.35,64.85,5,0.0
2024-03-25,ORD-2023-00002,C0254,Isidora Rojas,Consumidor,Concepción,Sur,P0087,Oficina,Papelería,Papelería Flex 7,629.31,115.33,1,0.15
2024-03-25,ORD-2023-00002,C0254,Isidora Rojas,Consumidor,Concepción,Sur,P0021,Tecnología,Computadores,Computadores Plus 1,1429.92,529.33,2,0.1
2024-02-20,ORD-2023-00003,C0203,Ignacio Araya,Corporativo,Valparaíso,Centro,P0093,Oficina,Archivadores,Archivadores Flex 3,3186.84,1333.8,6,0.15
2024-02-20,ORD-2023-00003,C0203,Ignacio Araya,Corporativo,Valparaíso,Centro,P0029,Tecnología,Computadores,Computadores Eco 9,787.14,-68.77,4,0.3
2024-02-20,ORD-2023-00003,C0203,Ignacio Araya,Corporativo,Valparaíso,Centro,P0034,Tecnología,Periféricos,Periféricos Flex 4,1007.23,179.29,3,0.2
2023-10-29,ORD-2023-00004,C0258,Fernanda Silva,Corporativo,Talca,Centro,P0089,Oficina,Papelería,Papelería Plus 9,2247.17,378.49,8,0.2


##6.**Justificaciones de codigo**


    
###**Justificación de negocio 1 - Información temporal:**
 Se incorporan la fecha y el identificador del pedido porque permiten analizar la evolución de las ventas en el tiempo y mantener la trazabilidad de cada transacción comercial.


---
###**Justificación de negocio 2 - Información del cliente:**
Se incorpora la información de los clientes para identificar quién realiza las compras y comparar el comportamiento comercial entre los diferentes segmentos


---
###**Justificación de negocio 3 - Información geográfica:**
Se incorpora la ubicación geográfica para identificar las ciudades y regiones con mayor desempeño comercial y apoyar la toma de decisiones.

---
###**Justificación de negocio 4 – Información del producto:**
Se incorpora la información de los productos para analizar su desempeño comercial y evaluar la rentabilidad por categoría y subcategoría.


---
###**Justificación de negocio 5 – Métricas comerciale:**
Se incorporan las métricas comerciales que permitirán evaluar el volumen de ventas, los ingresos, la rentabilidad y el impacto de los descuentos.




##7.**Justificación de los INNER JOIN**

INNER JOIN 1 – **Detalle_Pedido -> Pedidos**
  
  Se relaciona la tabla Detalle_Pedido con Pedidos para incorporar la información temporal de cada venta mediante la fecha del pedido. Además, esta relación permite acceder a la información del cliente, que será necesaria para analizar el comportamiento de compra y continuar la integración del Dataset Maestro.

---
INNER JOIN 2 – **Pedidos -> Clientes**

Se relaciona la tabla Pedidos con Clientes para identificar quién realizó cada compra. Esta integración permitirá analizar el comportamiento de los clientes, segmentar sus compras e identificar aquellos que generan mayores ingresos para la empresa.

---
INNER JOIN 3 – **Clientes -> Cliente_Ubicacion**

Se relaciona la tabla Clientes con Cliente_Ubicacion para asociar cada cliente con su ubicación geográfica y continuar el proceso de integración hacia la información territorial.

---
INNER JOIN 4 – **Cliente_Ubicacion -> Geografia**

Se relaciona la tabla Cliente_Ubicacion con Geografia para incorporar la ciudad y la región de cada cliente, permitiendo realizar análisis geográficos del comportamiento comercial.

---
INNER JOIN 5 – **Detalle_Pedido -> Productos**  

Se relaciona la tabla Detalle_Pedido con Productos para complementar las métricas comerciales con la información de cada producto, permitiendo analizar su desempeño y la rentabilidad por categoría y subcategoría.





##8.**Validación del Dataset Maestro**

###**Objetivo:** El propósito de esta etapa es verificar que el Dataset Maestro fue construido correctamente y que la información integrada mantiene la consistencia necesaria para responder las preguntas de negocio definidas en el proyecto.

###Validación 1 — **Cantidad de registros**

Primero queremos responder:

¿Cuántos registros comerciales contiene el Dataset Maestro?

In [63]:
%%sql
SELECT COUNT(*) AS Total_Registros
FROM Detalle_Pedido;

 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


Total_Registros
3026


**Justificación de negocio**

Se verificó que el Dataset Maestro está compuesto por 3026 registros comerciales, confirmando que la información fue integrada correctamente desde la tabla principal y se encuentra disponible para el análisis de negocio.


###Validación 2 — **Vista previa del Dataset**


**Justificación**

Se realizó una inspección visual de las primeras filas del Dataset Maestro para verificar que la integración entre pedidos, clientes, productos y ubicación geográfica se realizó correctamente antes de continuar con el análisis.


###Validación 3 — **Verificación de valores nulos**

In [54]:
%%sql
SELECT
COUNT(*) AS Registros_Sin_Fecha
FROM Pedidos
WHERE Fecha_Pedido IS NULL;


 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


Registros_Sin_Fecha
0


In [57]:
%%sql
SELECT
COUNT(*) AS Registros_Sin_Venta
FROM Detalle_Pedido
WHERE Ventas IS NULL;

 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


Registros_Sin_Venta
0


In [64]:
%%sql
SELECT
COUNT(*) AS Registros_Sin_Ganancias
FROM Detalle_Pedido
WHERE Ganancia IS NULL;

 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


Registros_Sin_Ganancias
0


**La justificación**

Se verificó que la columna Ventas,fecha y gancias no presentan valores nulos, garantizando que la principales métricas comerciales del proyecto se encuentra completas y listas para el análisis.

### Validación 5 -- **Duplicados**

In [62]:
%%sql
SELECT
ID_Pedido,
ID_Producto,
COUNT(*) AS Total
FROM Detalle_Pedido
GROUP BY
ID_Producto,
ID_Pedido
HAVING COUNT(*) > 1;

 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


ID_Pedido,ID_Producto,Total


##8.**Creación de la primera métrica calculada-(CASE WHEN)**

In [79]:
%%sql

SELECT

c.ID_Cliente,

c.Nombre_Cliente,

SUM(dp.Ventas) AS Total_Ventas,

CASE

WHEN SUM(dp.Ventas) >= 10000 THEN 'Cliente VIP'

WHEN SUM(dp.Ventas) >= 5000  THEN 'Cliente Frecuente'

ELSE 'Cliente Ocasional'

END AS Clasificacion_Clientes

FROM Detalle_Pedido dp

INNER JOIN Pedidos  p
ON dp.ID_Pedido = p.ID_Pedido

INNER JOIN Clientes c
ON p.ID_Cliente = c.ID_Cliente

GROUP BY

c.ID_Cliente,
c.Nombre_Cliente;

 * sqlite:////34_SuperTienda_Espanol.db
   sqlite:///34_SuperTienda_EspanolSuperTienda_Espanol.db
   sqlite:///SuperTienda_Espanol.db
Done.


ID_Cliente,Nombre_Cliente,Total_Ventas,Clasificacion_Clientes
C0001,Fernanda Díaz,3908.3700000000003,Cliente Ocasional
C0002,Valentina Muñoz,20655.07,Cliente VIP
C0003,Diego Castillo,7005.699999999999,Cliente Frecuente
C0004,Nicolás Silva,25636.410000000003,Cliente VIP
C0005,Daniela Rodríguez,27152.269999999997,Cliente VIP
C0006,Vicente Rojas,30261.860000000004,Cliente VIP
C0007,Daniela López,22823.71,Cliente VIP
C0008,Jorge Muñoz,36040.740000000005,Cliente VIP
C0009,Isidora Morales,16642.91,Cliente VIP
C0010,Valentina Pérez,22068.18,Cliente VIP


**Justificación de negocio**

Se creó una variable calculada utilizando CASE WHEN para clasificar a los clientes según el total de ventas generadas. Esta métrica permite identificar distintos niveles de aporte económico al negocio y facilita la definición de estrategias comerciales orientadas a la fidelización y retención de clientes.

---
**Justificación de negocio**

Se relaciona la tabla **Detalle_Pedido** con **Pedidos** para acceder a la información del cliente asociada a cada transacción. Esta relación permite conectar las ventas registradas con el cliente que realizó la compra, siendo un paso indispensable para calcular el total de ventas por cliente.

---
**Justificación de negocio**

Se relaciona la tabla **Pedidos** con **Clientes** para identificar a cada comprador mediante su nombre e identificador. Esta integración permite agrupar las ventas por cliente y construir una clasificación basada en su aporte económico al negocio.

